src/main/java/frc/robot/hardware/phoenix6/signal/Phoenix6ThreadAngleSignal.java

```diff
@@ -0,0 +1,48 @@
+package frc.robot.hardware.phoenix6.signal;
+
+import com.ctre.phoenix6.StatusSignal;
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.signal.ArrayAngleSignal;
+import frc.robot.subsystems.swerve.OdometryThread;
+import frc.utils.AngleUnit;
+import frc.utils.TimedValue;
+
+import java.util.ArrayList;
+import java.util.Queue;
+
+public class Phoenix6ThreadAngleSignal extends ArrayAngleSignal {
+
+	private final Queue<TimedValue<Double>> threadTimedValues;
+	private final OdometryThread thread;
+	private final StatusSignal<?> statusSignal;
+
+	protected Phoenix6ThreadAngleSignal(StatusSignal<?> statusSignal, AngleUnit angleUnit, OdometryThread thread) {
+		super(statusSignal.getName(), angleUnit);
+		this.thread = thread;
+		this.statusSignal = statusSignal;
+		this.threadTimedValues = thread.addSignal(statusSignal);
+	}
+
+	public void addWithLatencyCompensation(Phoenix6ThreadAngleSignal slopeSignal) {
+		thread.addLatencyAndSlopeSignals(statusSignal, threadTimedValues, slopeSignal.getStatusSignal());
+	}
+
+	public StatusSignal<?> getStatusSignal() {
+		return statusSignal;
+	}
+
+	@Override
+	protected ArrayList<TimedValue<Rotation2d>> updateValues(ArrayList<TimedValue<Rotation2d>> timedValues) {
+		thread.ThreadQueuesLock.lock();
+		try {
+			timedValues.clear();
+			threadTimedValues.forEach(
+				(timedValue) -> timedValues.add(new TimedValue<>(angleUnit.toRotation2d(timedValue.getValue()), timedValue.getTimestamp()))
+			);
```

Comment:
```suggestion
			timedValues.addAll(
				threadTimedValues.stream()
					.map(timedValue -> new TimedValue<Rotation2d>(angleUnit.toRotation2d(timedValue.getValue()), timedValue.getTimestamp()))
					.toList()
			);
```
this copies the array only once